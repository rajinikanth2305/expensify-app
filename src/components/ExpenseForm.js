import React from "react"


import moment from "moment"
import { SingleDatePicker } from "react-dates";
//set up note state 
//const date=new Date();
//moment is standard for dates
const now = moment();
console.log(now.format("MMM Do,YYYY"))
//on change and value for text area
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            createAt: props.expense ? moment(props.expense.createAt) : now,
            calanderFocused: false,
            error: ""


        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => (
            {
                description
            }
        )
        )
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => (
            {
                note
            })
        )
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(
                {
                    amount
                }
            )
        }
    };
    onDateChange = (createAt) => {
        if (createAt) {
            this.setState({
                createAt
            })

        }
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            //set error message please provide description and amount
            this.setState(
                {
                    error: "please add description and amount"
                }
            )

        }
        else {
            this.setState(() => (
                {
                    error: ""
                }
            )
            )
            this.props.onSubmit(
                {
                    description: this.state.description,
                    amount: parseFloat(this.state.amount, 10) * 100,
                    createAt: this.state.createAt.valueOf(),
                    note: this.state.note
                }
            )
        }
    }

    render() {
        return (
            <form className="form"onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}

                <input
                    className="text-input"
                    type="text"
                    placeholder="description"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    autoFocus />

                <input
                    type="text"
                    className="text-input"

                    placeholder="amount"
                    value={this.state.amount}

                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calanderFocused}
                    onFocusChange={({ focused }) => this.setState({ calanderFocused: focused })}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className="textarea"
                    palceholder="add a note for the expense(optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}>

                </textarea>
                <div>
                <button className="button">Add expense</button>
                </div>
            </form>

        )

    }
}
//refer https://www.npmjs.com/package/react-dates for singledatepicker
//moment is used for to get an date instance and this is the best and better one
//refer react-dates
//we imported react-dates and moment library 
//moment() fucntion gives the current tiem t
//we are passing timestamp to moment() to get the date where the expense was created
//we can simulate the events and we can pass the function to that simulate and also we can set object for that event value
