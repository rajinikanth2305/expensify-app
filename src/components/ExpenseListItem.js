import React from "react";
import {NavLink} from "react-router-dom";
import moment from "moment"
import numeral from "numeral"
//export stateless functional component 
//render few pieces of information 
//description 
//amount
//creat//
const ExpenseListItem=({id,description,amount,createAt})=>(
    <div>
    <NavLink to ={`/edit/${id}`} exact={true}>
    <h1>{description}</h1>
    </NavLink>
    <p>
    {numeral(amount/100).format('$0,0.00')}
    -
    {moment(createAt).format("MMMM Do,YYYY")}</p>
    <p>{id}</p>
    </div>
 
)
export default ExpenseListItem;


//notes for this lesson

//we should have to pass expense id to edit expens epage to get edit 
//we can pass using {`/edit/${id}`} to pass id to that expense page
