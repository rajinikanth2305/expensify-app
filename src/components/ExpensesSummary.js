import React from "react"
import selectTotalExpenses from "../selectors/expense-total"
import selectExpenses from "../selectors/expenses"
import { connect } from "react-redux"
import numeral from "numeral"

export const ViewTotalExpenses=({total,length})=>
{
    const expenseWord= length===1? "expense": "expenses";
    const formattedExpenseTotal=numeral(total/100).format(('$0,0.00'));
    return(
        <div>
        <h1>Viewing {length} {expenseWord} totalling is {formattedExpenseTotal}</h1>
        
        </div>
    )
}

const mapStateToProps=(state) => {
    const visibleExpenses=selectExpenses(state.expenses,state.filters)
    return {
        total:selectTotalExpenses(visibleExpenses),
        length:visibleExpenses.length
    };
  }
export default connect(mapStateToProps)(ViewTotalExpenses)
//export default ViewTotalExpenses;