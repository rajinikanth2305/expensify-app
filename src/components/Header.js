import React from "react";
import {NavLink} from "react-router-dom";

const Header=()=>
(
    <header>
    <h1>Tilte of tha expensify app</h1>
    <NavLink to ="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
    <NavLink to ="/create" activeClassName="is-active">create expense page</NavLink>
    <NavLink to="/edit" activeClassName="is-active">Edit page</NavLink>
    <NavLink to="/help" activeClassName="is-active">Help</NavLink>

    </header>
)
export default Header;