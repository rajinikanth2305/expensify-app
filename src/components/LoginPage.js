import React from "react"
import {BrowserRouter,Route,Switch,Link,NavLink} from "react-router-dom"
import {connect} from "react-redux"
import {startLogin} from "../actions/auth";
export const LoginPage=({startLogin})=>
{
    return(
        <div>
        <center>
        <button onClick={startLogin}>
        Login
        </button>
        </center>
        </div>
    )
}

const mapDispatchToProps=(dispatch)=>(
    {
        startLogin:()=>dispatch(startLogin())
    }
)
export default connect(undefined,mapDispatchToProps)(LoginPage);