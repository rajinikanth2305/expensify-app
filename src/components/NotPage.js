import React from "react";
import {Link} from "react-router-dom";
const NotPage=()=>(
    <div>
    
    This is not found help page-<Link to="/">Home</Link>
    </div>

)
export default NotPage