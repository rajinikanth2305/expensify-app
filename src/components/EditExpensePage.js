import React from "react";
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {editExpense} from "../actions/expenses"
import {removeExpense,startEditExpense,startRemoveExpense} from "../actions/expenses"

//refractor edit expense page to be class compoenent based
//setup mapdispatch to props,editexpense and removeexpense
//should render editexpensepage
//snapshot
//shouuld handle editexpense
//spy

//should handle removeexpense
//spy
export class EditExpensePage extends React.Component{
  onSubmit=(expense)=>
  {
    this.props.startEditExpense(this.props.expense.id,expense);
    this.props.history.push("/");
  }
  onClick=()=>
  {
    const id=this.props.expense.id;
    this.props.startRemoveExpense({id});
    this.props.history.push("/")
  }
  render()
  {
    return(
      <div>
      <div className="page-header">
      <div className="content-container">
      <h1 className="page-header__title">Edit Expense</h1>
      
      </div>
      </div>
      <div className="content-container">
      <ExpenseForm onSubmit={this.onSubmit}
      expense={this.props.expense}
      />
      <button className="button button--secondary" onClick={this.onClick}>Remove</button>

      </div>
   
      </div>

    )
  }

}
const mapStateToProps = (state, props) => {
    return {
      expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
  };
const mapDispatchToProps=(dispatch,props)=>
(
  {
    startEditExpense:(id,expense)=>dispatch(startEditExpense(id,expense)),
    startRemoveExpense:(id)=>dispatch(startRemoveExpense(id))
  }
)
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);

//we can pass the current component props to mapStateToprops passing props as argument
//props.history.push is for to go to homepage
//