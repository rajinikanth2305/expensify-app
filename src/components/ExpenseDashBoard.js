import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListItem from "./ExpenseListItem"
import ExpenseListFilters from "./ExpenseListFilter"
import ViewTotalExpenses from './ExpensesSummary'
const ExpenseDashboardPage = () => (
  <div>
    <ViewTotalExpenses />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
