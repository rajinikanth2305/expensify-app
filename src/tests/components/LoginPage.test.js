import React from 'react';
import { shallow } from 'enzyme';
import {LoginPage} from '../../components/LoginPage';

test('should render Login Page correctly', () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot();
});

test("should call startLogin on button click",()=>
{
  const startLogin=jest.fn();
  const wrapper=shallow(<LoginPage startLogin={startLogin} />);
  wrapper.find("button").simulate("click");
  expect(startLogin).toHaveBeenCalled();


})




//snapshot allows to track changes in data
//renderer function takes render function and it prints the output for component
//we have to install ReactShallowRender for testing the components
//library called nyzyme  for airbnb to actually writing test cases
//step 1 to install  