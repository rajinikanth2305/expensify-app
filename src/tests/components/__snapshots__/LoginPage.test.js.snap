// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login Page correctly 1`] = `
<div>
  <center>
    <button>
      Login
    </button>
  </center>
</div>
`;
